types/tapable/index.d.ts

```diff
@@ -210,4 +227,63 @@ declare namespace Tapable {
     }
 }
 
-export = Tapable
+type TapType = "sync" | "async" | "promise";
+
+export interface HookCompileOptions {
+    type: TapType;
+}
+
+export interface Tap {
+    name: string;
+    type: TapType;
+    fn: Function;
+    stage: number;
+    context: boolean;
+}
+
+export class Hook<T1 = any, T2 = any, T3 = any> {
```

Comment:
why you need to default to `any`?

Reply:
I'm definitely looking for suggestions here because I think I might have gotten in over my head a bit, but here's my line of reasoning on this:

Take this set of hooks as a sample: https://github.com/webpack/webpack/blob/8a59ef79b3124f76cbc6b19f19b2276f06178b6d/lib/Compiler.js#L29
```javascript
shouldEmit: new SyncBailHook(["compilation"]),
done: new AsyncSeriesHook(["stats"]),
additionalPass: new AsyncSeriesHook([]),
beforeRun: new AsyncSeriesHook(["compilation"]),
run: new AsyncSeriesHook(["compilation"]),
emit: new AsyncSeriesHook(["compilation"]),
afterEmit: new AsyncSeriesHook(["compilation"]),
thisCompilation: new SyncHook(["compilation", "params"]),
compilation: new SyncHook(["compilation", "params"]),
normalModuleFactory: new SyncHook(["normalModuleFactory"]),
contextModuleFactory: new SyncHook(["contextModulefactory"]),
beforeCompile: new AsyncSeriesHook(["params"]),
compile: new SyncHook(["params"]),
make: new AsyncParallelHook(["compilation"]),
afterCompile: new AsyncSeriesHook(["compilation"]),
watchRun: new AsyncSeriesHook(["compiler"]),
failed: new SyncHook(["error"]),
invalid: new SyncHook(["filename", "changeTime"]),
watchClose: new SyncHook([]),
```
You can see that each hook accepts a different number of parameters... (The max I've seen in the source has been 3, I figured I would start with that many..

I wanted a way to strongly type the callback fns for these hooks, but from what I've been reading theres no way to do this right now.
 (https://github.com/Microsoft/TypeScript/issues/5453)

By defining Hook as I have, I was able to define these hooks like so:

```typescript
additionalPass: AsyncSeriesHook;
beforeRun: AsyncSeriesHook<Compilation>;
run: AsyncSeriesHook<Compilation>;
emit: AsyncSeriesHook<Compilation>;
afterEmit: AsyncSeriesHook<Compilation>;
thisCompilation: SyncHook<Compilation, { normalModuleFactory: NormalModuleFactory }>;
compilation: SyncHook<Compilation, { normalModuleFactory: NormalModuleFactory }>;
normalModuleFactory: SyncHook<NormalModuleFactory>;
```

And then when I tap the hook, the parameters are strongly typed. In this example:
```
compiler.hooks.compilation.tap(
    "DllEntryPlugin",
    (compilation, { normalModuleFactory }) => {
        const dllModuleFactory = new DllModuleFactory();
        compilation.dependencyFactories.set(
            DllEntryDependency,
            dllModuleFactory
        );
    }
);
```

The callback signature is equivalent to:
```typescript
(compilation: webpack.compilation.Compilation, { normalModuleFactory: webpack.compilation.NormalModuleFactory })
```

I really hope that makes sense... I'm kinda struggling with this.

Reply:
thanks for explaining. Makes sense

Overall I feel Webpack source is a good candidate for TypeScriptification. Not sure how open @sokra and @TheLarkInn are to this idea! I love to help type the code.

Reply:
That would be the most ideal thing!